src/core/tasks_pool.c

```diff
@@ -0,0 +1,137 @@
+#include "tasks_pool.h"
+
+xccl_status_t tasks_pool_init(xccl_progress_queue_t* handle) {
```

Comment:
general comment, i personally don't really like name 'task_pool' since what we have here is NOT a pool of tasks. This is the lock_free implementation of progress queue.

@lappazos The API interface you added is not exactly what i meant. You kept xccl_progress_queue as is and added the API for progress_queue_enqueue/dequeue. I thought the the API would abstract progress_queue_init/do_progress/cleanup. In a single threaded case we don't need pop/push interface at all. We can only have ucs list and loop through it (example: https://github.com/vspetrov/mccl/blob/7fc9a6cfcc9f0f59184c761928ee6755b1edfe3e/src/core/xccl_progress_queue.c#L8 - possible implementation for single thread under #if ST == 1). 

It the MT case we go with your implementation and do pop/posh.

Reply:
name changed. This is the only possible API i know as long as it's a running time variable, and not environment. 
Regarding UCS list - i believe the data structure i have used eventually is a little more efficient, please take a look